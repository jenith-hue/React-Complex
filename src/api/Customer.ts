import { CancelToken } from 'axios';
import { appConfig } from '../config/app-config';
import { FilterPayload } from '../domain/PastDueList/Filter';
import { TextTemplate } from '../domain/TextConversation/TextTemplate';
import { LogWorkedHistoryPayload } from '../domain/WorkedHistory/WorkedHistory';
import {
  ActivityPayload,
  TakeCommitmentInput,
  AgreementInfoResponse,
  AssignRoutePayload,
  CustomerDetailsResponse,
  FieldSheetResponse,
  GetCustomerFieldSheetsPayload,
  StoreRouteResponse,
  WorkedHistoryResponse,
  CommunicationsDetailsResponse,
  CustomerPhoneInstructionPayload,
} from '../types/types';
import { client } from './client';

export const getPastDueList = async (
  payload: FilterPayload,
  offset: number,
  limit: number,
  cancelToken?: CancelToken
) =>
  client(`past-due-list?limit=${limit}&offset=${offset}`, {
    method: 'POST',
    body: payload,
    cancelToken,
  });

export const getCustomerDetails = (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  customerId: string,
  cancelToken?: CancelToken
): Promise<CustomerDetailsResponse> => {
  return client(`customers/${customerId}/details`, {
    method: 'GET',
    cancelToken,
  });
};

export const getWorkedHistory = (
  storeNumber: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  customerId: string,
  cancelToken: CancelToken
): Promise<WorkedHistoryResponse> =>
  client(`stores/${storeNumber}/customers/${customerId}/work-history`, {
    method: 'GET',
    cancelToken,
  });

export const getStoreRoutes = (
  storeNumber: string,
  cancelToken?: CancelToken
): Promise<StoreRouteResponse> =>
  client(`store-routes/${storeNumber}`, {
    method: 'GET',
    cancelToken,
  });

export const logWorkedHistory = (
  payload: Partial<LogWorkedHistoryPayload>
): Promise<WorkedHistoryResponse> =>
  client('customer/logactivity', {
    method: 'POST',
    body: payload,
  });

export const getAgreementInfo = (
  customerId: string,
  cancelToken?: CancelToken
): Promise<AgreementInfoResponse> =>
  client(`customer/get-agr-pastdue/${customerId}`, {
    method: 'GET',
    cancelToken,
  });

export const getTextTemplates = (): Promise<TextTemplate[]> =>
  client('/text-message/get-text-templates', { method: 'GET' });

export const logActivity = (payload: ActivityPayload) =>
  client('/customer/logactivity', { method: 'POST', body: { ...payload } });

export const assignRoute = (payload: AssignRoutePayload) =>
  client('/customer/assignroute', { method: 'POST', body: { ...payload } });

export const getFieldSheets = (
  payload: GetCustomerFieldSheetsPayload
): Promise<FieldSheetResponse> =>
  client('customers/field-sheets', { method: 'POST', body: { ...payload } });

export const takeCommitment = (payload: TakeCommitmentInput) =>
  client('/customer/take-commitment', { method: 'POST', body: { ...payload } });

export const getCommunicationDetails = (
  customerId: string
): Promise<CommunicationsDetailsResponse> =>
  client(`/customers/${customerId}/communication-details`, { method: 'GET' });

export const addressValidator = (payload: any) =>
  client(
    'customer/address/validate',
    {
      method: 'POST',
      body: {
        ...payload,
      },
    },
    appConfig.apiUrls.customer
  );

export const updatePhoneInstructions = (
  payload: CustomerPhoneInstructionPayload
) => client('/customers', { method: 'PUT', body: { ...payload } });

export const updateCustomer = (payload: any) =>
  client('customers', {
    method: 'PUT',
    body: {
      ...payload,
    },
  });

export const getCustomerStatesList = () =>
  client('references/entity?referenceCode=state_province', {
    method: 'GET',
  });
