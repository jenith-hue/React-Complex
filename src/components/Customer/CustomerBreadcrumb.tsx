import {
  makeStyles,
  RACBreadcrumb,
  RACBreadcrumbLinkType,
} from '@rentacenter/racstrap';
import clsx from 'clsx';
import React, { useEffect } from 'react';
import { AppRoute } from '../../config/route-config';
import { useCustomerDetails } from '../../context/CustomerDetails/CustomerDetailsProvider';
import { PastDueCustomerResponse } from '../../domain/PastDueList/PastDueCustomerList';
export interface CustomerProps {
  customer?: PastDueCustomerResponse;
}

const useStyles = makeStyles((theme: any) => ({
  breadcrumbContainer: {
    width: '100%',
    height: theme.typography.pxToRem(45),
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
}));

export const initialLinks = [
  {
    href: `/am${AppRoute.Index}`,
    label: 'Dashboard',
  },
  {
    href: `/am${AppRoute.PastDueList}`,
    label: 'Past Due List',
  },
];

export const initialLinksM = [
  {
    href: `/menu${AppRoute.Dashboard}`,
    label: 'Dashboard',
  },
  {
    href: `/menu${AppRoute.PastDueList}`,
    label: 'Past Due List',
  },
];

const getLinks = () => {
  if (window.location.pathname.includes('menu')) {
    return initialLinksM;
  }
  return initialLinks;
};
const getCustomerPath = (customerId: any) => {
  if (window.location.pathname.includes('menu')) {
    return `/menu/customer/update/${customerId}/customer`;
  }
  return 'am/customer';
};

export const CustomerBreadcrumb = () => {
  const [links, setLinks] =
    React.useState<RACBreadcrumbLinkType[]>(initialLinks);
  const { customerDetails } = useCustomerDetails();
  const { firstName, lastName, customerId } = customerDetails || {};
  const classes = useStyles();

  useEffect(() => {
    if (!customerDetails) return;
    const links = getLinks();
    const label =
      firstName && lastName && customerId
        ? `${firstName || ''} ${lastName || ''} - ${customerId || ''}`
        : '';
    setLinks([
      ...links,
      {
        href: `${getCustomerPath(customerId)}`,
        label: label,
      },
    ]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [customerDetails]);

  return (
    <div className={clsx(classes.breadcrumbContainer)}>
      <RACBreadcrumb links={links} />
    </div>
  );
};
