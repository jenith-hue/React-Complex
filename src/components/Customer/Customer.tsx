import React from 'react';
import { PastDueCustomerResponse } from '../../domain/PastDueList/PastDueCustomerList';
import { CustomerBreadcrumb } from './CustomerBreadcrumb';
import { CustomerDetailsProvider } from '../../context/CustomerDetails/CustomerDetailsProvider';
import { CustomerDetails } from './CustomerDetails/CustomerDetails';
import { Redirect, useLocation } from 'react-router-dom';
import { CustomerLocationState } from '../../types/types';
import { AppRoute } from '../../config/route-config';

export interface CustomerProps {
  customer?: PastDueCustomerResponse;
}
export const Customer = () => {
  const location = useLocation<CustomerLocationState>();

  const customerId = location?.pathname?.split('/')[3];

  if (!location?.state?.customer?.customerId && !customerId) {
    return <Redirect to={AppRoute.PastDueList} />;
  }
  return (
    <>
      <CustomerDetailsProvider>
        <CustomerBreadcrumb />
        <CustomerDetails />
      </CustomerDetailsProvider>
    </>
  );
};
