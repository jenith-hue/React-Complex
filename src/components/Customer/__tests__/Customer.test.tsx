/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { Customer } from '../Customer';

interface PropWithChildren {
  children: JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Grid: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACGrid" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../ContactInformation/ContactInformation', () => ({
  ContactInformation: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="ContactInformation" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../CustomerBreadcrumb', () => ({
  CustomerBreadcrumb: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerBreadcrumb" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../CustomerInformation/CustomerInformation', () => ({
  CustomerInformation: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerInformation" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../../../context/CustomerDetails/CustomerDetailsProvider', () => ({
  CustomerDetailsProvider: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerDetailsProvider" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../CustomerInformation/CustomerHeader', () => ({
  CustomerHeader: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerHeader" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../History/History', () => ({
  History: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="History" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../TextConversation/TextConversation', () => ({
  TextConversation: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="TextConversation" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock(
  '../../../context/TextConversationProvider/TextConversationProvider',
  () => ({
    TextConversationProvider: ({ children, ...rest }: PropWithChildren) => (
      <div data-testid="TextConversationProvider" {...rest}>
        {children}
      </div>
    ),
  })
);

jest.mock('../../../common/Footer/Footer', () => ({
  Footer: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Footer" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../CustomerInformation/CustomerInformationFooter', () => ({
  CustomerInformationFooter: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerInformationFooter" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../../../context/PrintLetter/PrintLetterProvider', () => ({
  PrintLetterProvider: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="PrintLetterProvider" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../../../context/WorkedHistory/WorkedHistoryProvider', () => ({
  WorkedHistoryProvider: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="WorkedHistoryProvider" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../../../context/CustomerHeader/CustomerHeaderProvider', () => ({
  CustomerHeaderProvider: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerHeaderProvider" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock('../CustomerDetails/CustomerDetails', () => ({
  CustomerDetails: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerDetails" {...rest}>
      {children}
    </div>
  ),
}));

// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '',
    state: {
      customer: {
        customerFirstName: 'First',
        customerLastName: 'Last',
        customerId: '123ID',
      },
    },
  }),
}));
describe('Customer', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Customer />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
