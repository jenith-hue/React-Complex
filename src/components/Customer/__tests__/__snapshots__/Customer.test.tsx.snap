// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer renders correctly 1`] = `
<div
  data-testid="CustomerDetailsProvider"
>
  <div
    data-testid="CustomerBreadcrumb"
  />
  <div
    data-testid="CustomerDetails"
  />
</div>
`;
